src/lib/sedlexing.ml

```diff
@@ -512,7 +590,45 @@ module Utf16 = struct
   end
 
   let from_gen s opt_bo = from_gen (Helper.gen_from_char_gen opt_bo s)
-  let from_channel ic opt_bo = from_gen (gen_of_channel ic) opt_bo
+
+  let from_channel ic opt_bo =
+    let t = Chan.create ic (chunk_size * 2) in
+    let bo = ref opt_bo in
+    let refill buf pos len =
+      let rec loop i =
+        if i = len then i
+        else (
+          match
+            let can_refill = i = 0 in
+            Chan.ensure t ~can_refill 2;
+            let c1 = Chan.get t 0 in
+            let c2 = Chan.get t 1 in
+            let o =
+              match !bo with
+                | Some o -> o
+                | None ->
+                    let o =
+                      match (Char.code c1, Char.code c2) with
+                        | 0xff, 0xfe -> Little_endian
+                        | _ -> Big_endian
+                    in
+                    bo := Some o;
+                    o
+            in
+            let c = Helper.number_of_char_pair o c1 c2 in
+            Chan.advance t 2;
+            Uchar.of_int c
```

Comment:
Looks like a good portion of this logic could be factored out to make it more readable. Something along these lines:

```
      let rec loop ~fetch i =
        if i = len then i
        else (
          match
            let can_refill = i = 0 in
            Chan.ensure t ~can_refill 1;
            let (c, n) = fetch t in
            Chan.advance t n;
            Uchar.of_char c
        )
      in
```
This would make the code more readable for anyone not familiar with it.

Reply:
I've done some more refactoring